In 428/inputs export

SaraBCL opened the pull request:
## Checklist

- [ ] Added entry to changelog (Skip)
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [X] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [Jira ticket](https://citizenlab.atlassian.net/browse/IN-428)

## What changes are in this PR?

Add a concise summary here of what happened, so the reviewer has some background.

## How urgent is a code review?

Not so urgent

SaraBCL committed:
Adds basic service

SaraBCL committed:
Adds endpoint

adessy commented on back/engines/commercial/insights/app/controllers/insights/web_api/v1/inputs_controller.rb:
```diff
@@ -3,7 +3,7 @@
   module WebApi::V1
     class InputsController < ::ApplicationController
-      skip_after_action :verify_policy_scoped, only: :index # The view is authorized instead.
+      skip_after_action :verify_policy_scoped, only: [:index, :index_xlsx] # The view is authorized instead.
```
We should add the :verify_authorized action to index and index_xlsx here:
```ruby
 after_action :verify_authorized
```

adessy commented on back/engines/commercial/insights/app/services/insights/xlsx_service.rb:
```diff
@@ -0,0 +1,30 @@
+module Insights
+	class XlsxService
+
+		def generate_inputs_xlsx(inputs, categories, view_private_attributes = false)
```
Watch out, a few tabs characters sneaked in 🙂

adessy replied:
```suggestion
		def generate_inputs_xlsx(inputs, categories, view_private_attributes: false)
```

Usually, we use keyword arguments for boolean arguments. It is also more consistent with `XlsxService`.

adessy commented on back/engines/commercial/insights/app/services/insights/xlsx_service.rb:
```diff
@@ -0,0 +1,30 @@
+      @xlsx_service ||= ::XlsxService.new
+    end
+
+	end
```
🚨 tab character

adessy commented on back/engines/commercial/insights/app/services/insights/xlsx_service.rb:
```diff
@@ -0,0 +1,30 @@
+      }
+
+      xlsx_service.generate_xlsx 'Inputs export', columns, categorized
+    end
```
(Just a suggestion -- mostly 💅 changes, the only difference I think is that it maintains the order of categories 

```suggestion
    def generate_inputs_xlsx(inputs, categories, view_private_attributes: false)
      columns = xlsx_service.generate_idea_xlsx_columns(inputs, view_private_attributes: view_private_attributes)

      category_cols = categories.map do |category|
        {
          header: category.name,
          f: lambda do |input|
            assignment = input.insights_category_assignments.find_by(category_id: category.id)
            return unless assignment

            assignment.approved ? 'approved' : 'suggested'
          end
        }
      end

      columns.insert(3, *category_cols)
      inputs = inputs.includes(:insights_category_assignments)
      xlsx_service.generate_xlsx('Inputs export', columns, inputs)
    end
```

adessy commented on back/engines/commercial/insights/app/controllers/insights/web_api/v1/inputs_controller.rb:
```diff
@@ -15,6 +15,19 @@ def index
+                                                                                         User).view_private_attributes?
+
+        send_data xlsx, type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+                        filename: 'inputs.xlsx'
```
(nice to have: the filename contains the name of the view (dasherized or snakecase))

SaraBCL replied:
A better filename is usually computed in the front-end, so skipping here (FE has more info on the context, this could be used in different scenarios in the future)

adessy commented on back/engines/commercial/insights/spec/acceptance/inputs_spec.rb:
```diff
@@ -67,7 +67,7 @@
         categories = create_list(:category, 2, view: view)
         bad_category = create(:category)
-        Insights::CategoryAssignmentsService.new.add_assignments_batch(ideas, categories.push(bad_category))
+        Insights::CategoryAssignmentsService.new.add_assignments_batch(ideas, categories| [bad_category])
```
I don't understand this change 🙂

adessy reviewed (approved):
Good to go! 💪 Just dropped a few comments as usual.

SaraBCL committed:
Apply suggestions from code review

Co-authored-by: adessy <adrien@citizenlab.co>

SaraBCL committed:
Review remarks fixes

SaraBCL committed:
Merge branch 'IN-428/inputs-export' of github.com:CitizenLabDotCo/citizenlab into IN-428/inputs-export

SaraBCL committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into IN-428/inputs-export

adessy committed:
IN-428 Rework index_xlsx

IN-428 #time 0w 0d 0h 1m Rework index_xlsx

adessy committed:
IN-428 Rubocop

IN-428 #time 0w 0d 0h 5m Rubocop

adessy committed:
IN-428 Rework inputs_spec.rb (cosmetic changes)

IN-428 #time 0w 0d 0h 15m Rework inputs_spec.rb (cosmetic changes)

adessy committed:
IN-428 Rework and fix xlsx_service_spec.rb (view_private_attributes is a keyword param)

IN-428 #time 0w 0d 0h 34m Rework and fix xlsx_service_spec.rb (view_private_attributes is a keyword param)

adessy committed:
Merge branch 'master' into IN-428/inputs-export

IN-428 #time 0w 0d 0h 5m Merge branch 'master' into IN-428/inputs-export
